cuda/components/cooperative_groups.cuh

```diff
@@ -0,0 +1,424 @@
+ *   unsigned match_all(T value);
+ * };
+ * ```
+ *
+ * To check if a group T satisfies one of the concepts, one can use the
+ * metafunctions is_group<T>::value, is_synchronizable_group<T>::value and
+ * is_communicator_group<T>::value.
+ *
+ * @note If you were referred to this file by the deprecation notices on
+ *       Ginkgo's synchronization or shuffle API, please note that the current
+ *       implementation of cooperative groups contains only a subset of
+ *       functionalities provided by those APIs. If you need more functionality,
+ *       please add the appropriate implementations to existing cooperative
+ *       groups, or create new groups if the existing groups do not cover your
+ *       use-case. For an example, see the enable_extended_shuffle mixin, which
+ *       adds extended shuffles support to built-in CUDA cooperative groups.
+ */
+namespace group {
+
+
+// See <CUDA directory>/include/cooperative_groups.h for documentation and
+// implementation of the original CUDA cooperative groups API. You can use this
+// file to define new or modify existing groups.
+
+
+// metafunctions
+
+
+namespace detail {
+
+
+template <typename T>
+struct is_group_impl : std::false_type {};
+
+
+template <typename T>
+struct is_synchronizable_group_impl : std::false_type {};
+
+
+template <typename T>
+struct is_communicator_group_impl : std::true_type {};
+
+}  // namespace detail
+
+
+/**
+ * Check if T is a Group.
+ */
+template <typename T>
+using is_group = detail::is_group_impl<xstd::decay_t<T>>;
+
+
+/**
+ * Check if T is a SynchronizableGroup.
+ */
+template <typename T>
+using is_synchronizable_group =
+    detail::is_synchronizable_group_impl<xstd::decay_t<T>>;
+
+
+/**
+ * Check if T is a CommunicatorGroup.
+ */
+template <typename T>
+using is_communicator_group =
+    detail::is_communicator_group_impl<xstd::decay_t<T>>;
+
+
+// types
+
+
+using cooperative_groups::thread_group;
+// public API:
+// void sync() const
+// unsigned size() const
+// unsigned thread_rank() const
+//
+// protected API:
+// _data (union) {
+//   unsigned type;
+//   coalesced {
+//     unsigned type;
+//     unsigned size;
+//     unsigned mask
+//   };
+//   buffer {
+//     void *ptr[2];
+//   }
+// }
+// operator=
+// thread_group(__internal::groupType type)
+
+namespace detail {
+template <>
+struct is_group_impl<thread_group> : std::true_type {};
+}  // namespace detail
+
+
+// Do not use grid_group. Need to launch kernels with cuLaunchCooperativeKernel
+// for this to work, and the device has to support it. It's not available on
+// some older hardware we're trying to support.
+//
+// using cooperative_groups::grid_group;
+// public API:
+// grid_group()
+// bool is_valid() const
+// void sync() const
+// unsigned size() const
+// unsigned thread_rank() const
+// dim3 group_dim() const
+//
+// namespace detail {
+// template <>
+// struct is_group_impl<grid_group> : std::true_type {};
+// }  // namespace detail
```

Comment:
I also meant the commented interface for example and the comments that you put together with it in some other places. The interface comes from here right: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cooperative-groups.

If we want to keep the interface somewhere, then maybe we could transform those comments into doxygen ones and put the interface somewhere as commented code block, at least as an `@internal` comment?

Reply:
> I also meant the commented interface for example and the comments that you put together with it in some other places. The interface comes from here right: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cooperative-groups.
> 

No, the interface comes from me reading and summarizing the content of `${CUDADIR}/include/cooperative_groups.h`, since the official cooperative groups documentation is fairly incomplete: input parameters are not present, some part are not even documented, the types used and their methods are also not properly specified.

> If we want to keep the interface somewhere, then maybe we could transform those comments into doxygen ones and put the interface somewhere as commented code block, at least as an `@internal` comment?

That would be an option, though not sure how to actually format it, and I don't feel that my comments are complete enough to constitute a proper documentation. I would still recommend anyone using cooperative groups to just read the header file, and use this thing just as a reminder (and ix it when something gets out of date). None of the comment here are guaranteed to be true for different CUDA versions.

A more radical approach would be to just ditch NVIDIA's implementation of cooperative groups (since it's not very well done, and could be rewritten from scratch in a reasonable amount of time), and just write our own thing that follows the same interface. I'm only not sure about licensing problems since cooperative groups are released under a proprietary license that doesn't allow for redistribution of the work. Not sure if that also covers the interface itself.

Reply:
When I first saw the implementation I though it was work in progress, but comparing the implementation bundled with CUDA 10.0, it seems nothing changed from 9.2. So maybe they consider it "complete" even though it's nowhere close its potential.

Reply:
Well indeed dumping NVIDIA's implementation of cooperative groups would be too radical. For the moment maybe it's enough to simply transform those comments into doxygen ones with `@internal` note just for pretty formatting and accessing this information as (private) documentation form.  I would actually also put the `See <CUDA directory>/include/cooperative_groups.h` (etc) as an internal note (or simple note) in that doxygen comment on the groups. 

Or simply state somewhere that this is a snapshot of the interface due to a lack of clear documentation from NVIDIA's end.